centralize reusable profiling code

chrismeyersfsu opened the pull request:
see doc updates for usage

Verified that `TimingMiddleware` continues to work.
![image](https://user-images.githubusercontent.com/722880/96455590-6e45ff80-11eb-11eb-85d8-b9b3acce53ea.png)

chrismeyersfsu committed:
centralize reusable profiling code

shanemcd commented:
recheck

ryanpetrello commented on awx/main/middleware.py:
```diff
@@ -32,45 +29,28 @@ class TimingMiddleware(threading.local, MiddlewareMixin):
+
     def process_request(self, request):
         self.start_time = time.time()
+        print(f"Setting is {settings.AWX_REQUEST_PROFILE}")
```
👃

ryanpetrello commented on awx/main/utils/profiling.py:
```diff
@@ -0,0 +1,148 @@
+            return None
+
+
+def profile(name, *init_args, **init_kwargs):
```
❤️

ryanpetrello commented on awx/main/middleware.py:
```diff
@@ -32,45 +29,28 @@ class TimingMiddleware(threading.local, MiddlewareMixin):
         if settings.AWX_REQUEST_PROFILE:
-            self.prof = cProfile.Profile()
-            self.prof.enable()
+            print("Starting it")
```
Do we need these print statements?

ryanpetrello commented on awx/main/middleware.py:
```diff
@@ -32,45 +29,28 @@ class TimingMiddleware(threading.local, MiddlewareMixin):
         return response
 
-    def save_profile_file(self, request):
-        if not os.path.isdir(self.dest):
```
Nice to lose all this.

ryanpetrello commented on awx/main/utils/profiling.py:
```diff
@@ -0,0 +1,148 @@
+import sys
+
+
+class AWXBase:
```
Better name, maybe `AWXProfileBase` ?

ryanpetrello commented on docs/debugging.md:
```diff
@@ -98,3 +98,81 @@ sdb-listen
+total 16
+drwxr-xr-x 2 awx  root 4096 Oct 15 13:25  .
+drwxrwxr-x 1 root root 4096 Oct 15 13:25  ..
+-rw-r--r-- 1 awx  root   74 Oct 15 13:25 '0:00:02.002159-seconds-my_task_manager_timing-e8c03054-7ea3-43ea-902e-b5f3561cbae7.time'
```
Why are there single quotes around this filename?

chrismeyersfsu replied:
Will double check. I think it's just my zsh terminal

ryanpetrello commented on docs/debugging.md:
```diff
@@ -98,3 +98,81 @@ sdb-listen
+total 24
+drwxr-xr-x 2 awx  root 4096 Oct 15 13:23 .
+drwxrwxr-x 1 root root 4096 Oct 15 13:23 ..
+-rw-r--r-- 1 awx  root  635 Oct 15 13:23 task_manager_profile-2.001s-2303-272858af-3bda-45ec-af9e-7067aa86e4f3.dot
```
One nice thing about the old format is that it _started_ with the timing, so you could sort the filenames with `ls -n` to see the slowest/fastest requests in order.  Might want to change this back to e.g.,

`2.00s-task_manager_profile-<UUID>.dot`

ryanpetrello reviewed (commented):
I like the way this is going.  Are you going to update the callback receiver to make use of this refactored code (instead of its own cProfile usage)?

chrismeyersfsu commented:
![image](https://user-images.githubusercontent.com/722880/96587001-c1818600-12af-11eb-91c0-33762471077e.png)
* confirmed, `'filename'` is because of my shell
* Moved duration to front of filename
* fixed Timing to output `total_seconds()` rather than `hh:mm:ss`
* removed debug print statements
* renamed base class
* replaced callback receiver profiler with this one

ryanpetrello commented on awx/main/middleware.py:
```diff
@@ -32,45 +29,25 @@ class TimingMiddleware(threading.local, MiddlewareMixin):
-            self.prof.disable()
-            cprofile_file = self.save_profile_file(request)
-            response['cprofile_file'] = cprofile_file
+            response['cprofile_file'] = self.prof.stop()
```
If we want to return this in the HTTP response, we should make this follow the convention of other custom headers, `X-API-Profile-File`

ryanpetrello reviewed (approved):
Thanks for cleaning up my mess.
